### DISCUSSION TITLE
Stop Grabbing Text Based On A Known Keyword

### DISCUSSION THREAD

**Discussion by HelloLudger** [peggyjs/peggy#279] [Q&A]
Hey,

I really love Peggy, but I can't get one thing to work. Imagine the following input:
```
## First
Hello World
Another Line

## Second
This text should be part of the second
variable, named s.
```

I want to create a grammar to get the content following the titles "## First" and "## Second". The result look something like this:
```
{
   "first": "Hello World
Another Line",
   "second": "This text should be part of the second
variable, named s."
}
```
The titles are fixed, always "## First" and "## Second", followed by a line break. We can assume for now, that they always exist.
I created the following grammar:
```
Block
  = f:First s:Second? br* { return {first: f, second: s} }

First
  = "## First" br l:Lines             { return l }
  
Second
  = "## Second" br l:Lines            { return l }

Lines
  = ls:Line*                           { return ls.join("\n").trim() }
  
Line
  = br                                 { return "" }
  / text:charw+ br?                    { return text.join("").trim() }

br "new line"
  = [\n\r]
  
charw "character with whitespace"
  = [^\n\r]
```
I made `s:Second` optional, because the output shows clearly the problem:
```
{
   "first": "Hello World
Another Line

## Second
This text should be part of the second
variable, named s.",
   "second": null
}
```

If it's not optional, the error is: `Line 7, column 19: Expected "## Second", character with whitespace, or new line but end of input found.`

I can't find a solution, even though I assume the problem is pretty common. 

Thank you in advance,
Ludger

**Comment by Mingun** [comment]
Your `Lines` rule consumes all input which you have. You should add a negative assertion to stop that. You need to check stop condition on each symbol, so:
```pegjs
charw "character with whitespace"
  = !Second @[^\n\r]
```
